### PR TITLE
Add `cargo config` subcommand.

### LINKED ISSUE #2362: Finish implementing `cargo config` and related support
Configuration in Cargo initially had a relatively grand vision, but much of it was unfortunately never implemented. This issue should help track finishing this support as well as explaining what it is intended to be!
### A `cargo config` subcommand

It was initially intended to have a `cargo config` subcommand ala `git config` or `bundle config`. Configuration values could be read/written through this interface to the various locations on the system as need be.

An example of this would be Bundler's [`bundle config`](http://bundler.io/v1.2/bundle_config.html) command.

The major blocker for this is being able to read a TOML file, modify some aspects of it, and then write it back to the filesystem. This should crucially preserve aspects such as comments, formatting, etc. Today we cannot do this.
### Environment variables for configuration

Right now we've got a [smattering of environment variables](http://doc.crates.io/environment-variables.html#environment-variables-cargo-reads), but it's somewhat inconsistent in terms of what's an environment variables, what's a command line option, and what's a configuration option. 

I believe it was originally intended that environment variables and configuration options should be one-and-the-same. All configuration options should have a mapping to an environment variable which provides the ability to read/write that field. For example [Bundler has this as well](http://bundler.io/v1.2/bundle_config.html#configuration_keys).

### PR DESCRIPTION
**Pull request by ehuss**
This adds an initial version of the `cargo config` command as discussed in #2362.

Closes #2362

### CODE REVIEW

**Comment by ehuss** [comment]
## Open questions

* Is there any preference in style for the `--merged` flag. Should it be `--merged=yes/no` or `--no-merged`?

## Future work

* Warnings for unknown keys aren't really handled. Not sure if this is possible.
* Quoted keys aren't handled as CLI input. For example, `config get 'target."cfg(target_os=\"linux\")".runner'`  Unfortunately toml doesn't expose the parsing functions to make this easy. Could do from_str("{} = 1", key), and then collect all the key components. I wouldn't want that in the normal Config.get() path, though, for performance.
* JSON doesn't show environment variables. Not sure how to handle that.
* The original design doc mentioned a few other output formats, but I only implemented 3.
* The design doc mentioned future subcommands like `set`, `delete`, and `edit`.

## Environment variable handling

I'm a little unhappy with the way environment variables are handled. If a command like `cargo config get profile.dev` is run, and the environment variable CARGO_PROFILE_DEV_OPT_LEVEL=3 is set, then it will only display the environment variable as a note in a comment instead of actually part of the value output. For json output, it is not displayed at all.

If getting a non-table value (like `cargo config get profile.dev.opt-level`), then it should work correctly and use the environment variable.

I'm not sure how to go about solving that. The only solution I can think of is defining some kind of schema that describes the structure of the config types. I can't think of a way to do this with serde.

Even with that, there are some "lazy" environment variables that I think would be impossible to handle correctly. They are:

* `CARGO_PROFILE_*_…`
* `CARGO_ALIAS_*`
* `CARGO_REGISTRIES_*_…`
* `CARGO_TARGET_*_…`

**Comment by Revantus** [comment]
Personally I prefer the `--no-merged` approach. Testing locally it seems that `--merged` is the default when using `cargo config get` which makes choosing `--merged` and `--merged yes` feel kind of redundant.

File: src/cargo/ops/cargo_config.rs [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+    );
+    for (env_key, env_value) in env {
+        let val = shell_escape::escape(Cow::Borrowed(env_value));
+        drop_println!(config, "# {}={}", env_key, val);
+    }
+}
+
+fn print_json(config: &Config, key: &ConfigKey, cv: &CV, include_key: bool) {
+    let json_value = if key.is_root() || !include_key {
+        match cv {
```
Comment (alexcrichton) [inline src/cargo/ops/cargo_config.rs:173]: Could this be deduplicated with `json_add` below where there's a function from `CV` to a json value?

**Review by alexcrichton — COMMENTED** [review]
>     Is there any preference in style for the --merged flag. Should it be --merged=yes/no or --no-merged?

Personally over the years I've come to appreciate `--flag=val` more than `--no-flag`, so I'd go for `--merged`. I don't feel strongly on this though.

> JSON doesn't show environment variables. Not sure how to handle that.

Since JSON goes to stdout, could warnings about this go to stderr perhaps?

> I'm not sure how to go about solving that. The only solution I can think of is defining some kind of schema that describes the structure of the config types. I can't think of a way to do this with serde.

Right now all of Cargo's configuration is entirely untyped and accessed on-the-fly so one of the issues with your `profile` example is that within Cargo we're always accessing profile configuration in a typed manner, but with this new subcommand it's instead all untyped so we don't know what to do with the map.

One possible alternative is we instead "register" configuration values with `Config` early on, so `Config` learns about the structure of all expected configuration. Basically your schema idea but with registration up-front of what the schema is. I have no idea how to do this in an ergonomic fashion though and it's definitely out of the scope of this PR.

**Comment by ehuss** [comment]
> Since JSON goes to stdout, could warnings about this go to stderr perhaps?

Yea, I felt a little uncomfortable with that, but I can't say why exactly.  I went ahead and pushed a commit that prints a note to stderr.

>  I have no idea how to do this in an ergonomic fashion though and it's definitely out of the scope of this PR.

Yea, I will try to think it over for a while.  I fear it will be a lot of code and changes, just to accommodate this one use case. I don't think it isn't super important right now.  My primary use case is just to help people debug configuration settings, and printing the environment variables is sufficient to me for that purpose.

**Pull request merged**